### ISSUE TITLE
Add support for Nerd Fonts (doublewidth icons)

### ISSUE THREAD

**Issue by powerman** [neovim/neovim#29496]
### Problem

**TL;DR:** It would be nice to patch tables in `src/unicode/` to mark codepoints used by Nerd Fonts doublewidth icons as doublewidth. Proof of concept (very likely implemented in wrong way/place but working): https://github.com/powerman/wcwidth-icons/blob/master/patches/neovim/wcwidth-icons.patch

Nerd Fonts today release fonts in few flavors (mentioned in font name's suffixes): default, "Propo" and "Mono":
```
NotoMonoNerdFont-Regular.ttf: NotoMono Nerd Font,NotoMono NF:style=Regular
NotoMonoNerdFontPropo-Regular.ttf: NotoMono Nerd Font Propo,NotoMono NFP:style=Regular
NotoMonoNerdFontMono-Regular.ttf: NotoMono Nerd Font Mono,NotoMono NFM:style=Regular
```

The "Mono" is designed for apps which does not support doublewidth glyphs. It will work everywhere, but icons will be too small and thus nearly unless - this is better than nothing but in no way fun!
`$ urxvt -fn 'xft:NotoMono Nerd Font Mono' -e nvim ~/nerd-symbols.txt`
![изображение](https://github.com/neovim/neovim/assets/1354301/8c962238-be8a-4e57-ae62-6abc4f49f662)

The other two flavors mostly don't work (only few icons are shown) because of missing support for doublewidth glyphs both in terminal and Neovim.
`$ urxvt -fn 'xft:NotoMono Nerd Font' -e nvim ~/nerd-symbols.txt`
![изображение](https://github.com/neovim/neovim/assets/1354301/195eeefa-3ea6-4ae3-a1ce-e4f9ecc6ce44)

If we'll use terminal with support for doublewidth glyphs the "Mono" will looks the same, but other two flavors will be broken: icons will be shown as doublewidth but as result it'll shown over next symbol (a space, in this case):
`$ wezterm --config "font = wezterm.font 'NotoMono Nerd Font Mono'" -e nvim ~/nerd-symbols.txt`
![изображение](https://github.com/neovim/neovim/assets/1354301/619dd360-4bef-4bb3-b6fb-764f274b2ad9)
`$ wezterm --config "font = wezterm.font 'NotoMono Nerd Font'" -e nvim ~/nerd-symbols.txt`
![изображение](https://github.com/neovim/neovim/assets/1354301/69c1d1ce-1a9b-4cb1-bc45-0061e980958f)
To make this more obvious we can open file with multiple icons in a row without spaces between them:
![изображение](https://github.com/neovim/neovim/assets/1354301/b68887f8-205f-4967-af36-b617fcad3c29)

There is a way to fix most old apps (e.g. urxvt) by using LD_PRELOAD hack to patch result of `wcwidth(3)` and `wcswidth(3)` to make them return 2 for doublewidth codepoints in Nerd Fonts. (One implementation is https://github.com/powerman/wcwidth-icons.) This way we can make icons work in urxvt, **but not in Neovim**:
`$ LD_PRELOAD=/usr/lib64/libwcwidth-icons.so urxvt -fn 'xft:NotoMono Nerd Font'`
![изображение](https://github.com/neovim/neovim/assets/1354301/e0ef30b4-49a6-40e3-8b36-3da360670eae)
![изображение](https://github.com/neovim/neovim/assets/1354301/f72ab13c-43ca-4964-996c-fefe04ff625a)

This is because some apps (vim and Neovim included) does not use `wcwidth(3)`, instead they hardcode list of doublewidth codepoints in their sources. So, to make them support nice large icons we need to patch these apps. Patch for vim is available at https://github.com/powerman/wcwidth-icons/blob/master/patches/vim/wcwidth-icons.patch. As for Neovim, today I've tried to implement such patch (and it works!), but current implementation feels like a dirty hack because to me it looks like this should be implemented in different way and place, which will likely require patching this logic:

https://github.com/neovim/neovim/blob/724d1110b1e4699a34f489e9cdb2d25098746499/src/nvim/generators/gen_unicode_tables.lua#L248-L251

Anyway, dirty hack or not, but it now works!
`$ LD_PRELOAD=/usr/lib64/libwcwidth-icons.so urxvt -fn 'xft:NotoMono Nerd Font' -e nvim ~/nerd-symbols.txt`
![изображение](https://github.com/neovim/neovim/assets/1354301/fd1c22ba-619a-43f1-8e33-0ab1f06b2a91)

### Expected behavior

This feature should be available in Neovim out-of-box, because a lot of Neovim users and plugins use Nerd Font icons.

**Comment by clason** [comment]
> t would be nice to patch tables in src/unicode/ to mark codepoints used by Nerd Fonts doublewidth icons as doublewidth.

We are trying to get _rid_ of those tables... see #26165

**Comment by powerman** [comment]
Cool! Just don't forget to test how utf8proc handles these doublewidth Nerd Fonts codepoints. :smile:

**Comment by powerman** [comment]
BTW, I've tried to use https://github.com/nvim-lua/kickstart.nvim and noticed it breaks with my patch:
![изображение](https://github.com/neovim/neovim/assets/1354301/0ddc2538-f95b-44c6-9123-69e43f93003b)

This happens because plugin author thinks Nerd Font icons are not doublewidth and adds extra space after an icon (it may not be obvious from the snippet, but if you'll try to copy&paste it you'll notice space after an icon inside a quoted string):

https://github.com/folke/todo-comments.nvim/blob/51e10f838e84b4756c16311d0b1ef0972c6482d2/lua/todo-comments/config.lua#L26-L28

Similar issues was already reported for some other plugins (e.g. https://github.com/puremourning/vimspector/issues/263) and was fixed by adding extra check for 'ambiwidth' width option: https://github.com/puremourning/vimspector/pull/264/files#diff-4607cd06061548fc2a8d9ee4fad6df8b1dc48d80dc905d9e68249f39134c257fR16-R17

**Comment by zeertzjq** [comment]
`setcellwidths()`
